File: 3217.delete-nodes-from-linked-list-present-in-array.py
#
# @lc app=leetcode id=3217 lang=python3
#
# [3217] Delete Nodes From Linked List Present in Array
#

# @lc code=start
# Definition for singly-linked list.
from typing import Optional, List
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def modifiedList(self, nums: List[int], head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = head
        cur = dummy
        nums = set(nums)
        while cur:
            while cur.next and cur.next.val in nums:
                cur.next = cur.next.next
            cur = cur.next
        return dummy.next
# @lc code=end

File: 3217.delete-nodes-from-linked-list-present-in-array.java
/*
 * @lc app=leetcode id=3217 lang=java
 *
 * [3217] Delete Nodes From Linked List Present in Array
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
import java.util.HashSet;
class Solution {
    public ListNode modifiedList(int[] nums, ListNode head) {
        // HashSet<Integer> set = new HashSet<>();
        // for(int num : nums){
        //     set.add(num); 
        // }
        // ListNode dummy = new ListNode();
        // dummy.next = head; 
        // ListNode cur = head;
        // ListNode prev = dummy;
        // while(cur != null){
        //     if(set.contains(cur.val)){
        //         prev.next = cur.next;
        //     }else{
        //         prev = cur;
        //     }
        //     cur = cur.next;

        // }
        // return dummy.next;


        // HashSet<Integer> set = new HashSet<>();
        // for(int num : nums) set.add(num);
        // ListNode dummy = new ListNode();
        // dummy.next = head;

        // ListNode cur = head;
        // ListNode prev = dummy;

        // while(cur != null){
        //     if(set.contains(cur.val)){
        //         prev.next = cur.next;
        //         cur = cur.next;
        //     }else{
        //         prev = cur;
        //         cur = cur.next;
        //     }
        // }
        // return dummy.next;

        HashSet<Integer> set = new HashSet<>();
        for(int num : nums) set.add(num);

        ListNode dummy = new ListNode();
        dummy.next= head; 
        ListNode cur = dummy;

        while(cur.next != null){
            ListNode next = cur.next;
            if(set.contains(next.val)){
                cur.next = next.next;
            }else{
                cur = next;
            }
        }
        return dummy.next;
    }
}
// @lc code=end

